src/test/groovy/graphql/normalized/ExecutableNormalizedOperationToAstCompilerTest.groovy

```diff
@@ -152,17 +149,650 @@ class ExecutableNormalizedOperationToAstCompilerTest extends Specification {
+
+        when:
+        def result = compileToDocument(schema, QUERY, null, tree.topLevelFields, noVariables)
+        def printed = AstPrinter.printAst(new AstSorter().sort(result.document))
+
+        then:
+        printed == """query {
+  animal {
+    ... on Cat {
+      __typename
+      name
+    }
+    ... on Dog {
+      isGoodBoy
+      name
+    }
+  }
+}
+"""
+    }
+
+    def "test typename in union when placed outside of fragments"() {
+        def schema = TestUtil.schema("""
+        union Animal = Cat | Dog
+        type Cat {
+            name: String
+            parent: Cat
+        }
+        type Dog {
+            name: String
+            parent: Dog
+            isGoodBoy: Boolean
+        }
+        type Query {
+            animal: Animal
+        }
+        """)
+
+        String query = """
+        {
+            animal {
+                __typename
+                ... on Cat {
+                    name
+                }
+                ... on Dog {
+                    name
+                }
+                ... on Dog {
+                    isGoodBoy
+                }
+            }
+        }
+        """
+
+        def tree = createNormalizedTree(schema, query)
+        // printTreeWithLevelInfo(tree, schema).forEach { println it }
+
+        when:
+        def result = compileToDocument(schema, QUERY, null, tree.topLevelFields, noVariables)
+        def printed = AstPrinter.printAst(new AstSorter().sort(result.document))
+
+        then:
+        printed == """query {
+  animal {
+    __typename
+    ... on Cat {
+      name
+    }
+    ... on Dog {
+      isGoodBoy
+      name
+    }
+  }
+}
+"""
+    }
+
+    def "test typename in union when placed inside fragments but on all types conditions"() {
+        def schema = TestUtil.schema("""
+        union Animal = Cat | Dog
+        type Cat {
+            name: String
+            parent: Cat
+        }
+        type Dog {
+            name: String
+            parent: Dog
+            isGoodBoy: Boolean
+        }
+        type Query {
+            animal: Animal
+        }
+        """)
+
+        String query = """
+        {
+            animal {
+                ... on Cat {
+                    __typename
+                    name
+                }
+                ... on Dog {
+                    __typename
+                    name
+                }
+                ... on Dog {
+                    isGoodBoy
+                }
+            }
+        }
+        """
+
+        def tree = createNormalizedTree(schema, query)
+        // printTreeWithLevelInfo(tree, schema).forEach { println it }
+
+        when:
+        def result = compileToDocument(schema, QUERY, null, tree.topLevelFields, noVariables)
+        def printed = AstPrinter.printAst(new AstSorter().sort(result.document))
+
+        then:
+        // Perhaps the typename should be hoisted out of the fragments, but the impl currently generates
```

Comment:
thinking about the __typename hanlding: I think we need to improve this
one fundamental principle of the ENF is that regardless of the query the ENF should be same if it executes the same.
Here we can generate two different ENOperation when we place __typename directly inside the Union or inside all type conditions for all types. Both of them must produce the same ENOperation. 

This is not directly related to printing, but in this test it becomes obvious.